### ISSUE TITLE
Linux Targets that require sudo

### ISSUE THREAD

**Issue by tsweaver321** [microsoft/vscode-makefile-tools#168]
How can a Linux target that requires sudo (e.g, scheduler changed to SCHED_FIFO) be launched and debugged with this extension?

**Comment by bobbrow** [comment]
I believe you would need to create a `launch.json` file for debugging ([more information here](https://code.visualstudio.com/docs/cpp/launch-json-reference)).  If desired, you can use the substitution variables [listed here](https://github.com/microsoft/vscode-cmake-tools/blob/develop/docs/cmake-settings.md#variable-substitution) to fill in the path to the target binary.

**Comment by gcampbell-msft** [comment]
@tsweaver321 Could you confirm that your question has been addressed?

We plan to close this issue if we don't hear from you, thanks!

**Comment by tsweaver321a** [comment]
> @tsweaver321 Could you confirm that your question has been addressed?
> 
> We plan to close this issue if we don't hear from you, thanks!

@gcampbell-msft No, I don't believe this addresses the issue of how to launch a debug session for a target that requires elevated permission (i.e., sudo) to run.  The first link is documentation on the launch.json, but it doesn't address sudo.  The 2nd link is broken, but by the description, it doesn't sound like it would address sudo either.

However, chatGPT provided a solution.  I haven't tried it yet, but it seems reasonable:

(ChatGPT):
To configure a debug session for a Linux target that requires `sudo` to run in Microsoft Visual Studio Code (VS Code), you can use the "Remote - SSH" extension in combination with the "C/C++" or another applicable extension for debugging. Here's a step-by-step guide:

1. **Install Required Extensions:**

   Make sure you have the following extensions installed in your VS Code:

   - [Remote - SSH](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh): This extension allows you to connect to a remote Linux machine over SSH.
   - An extension for the programming language you are debugging (e.g., "C/C++" for C/C++ debugging).

2. **Connect to the Remote Linux Machine:**

   Open VS Code and click on the green remote icon in the bottom-left corner. It should say "SSH: Connect to Host..." when you hover over it. If you don't see this icon, you can open the command palette (Ctrl+Shift+P) and type "Remote-SSH: Connect to Host" to initiate the connection.

   You'll be prompted to enter the SSH connection information for your Linux machine.

3. **Set up SSH Authentication:**

   You may need to configure your SSH keys or provide a password to connect to the remote machine.

4. **Open Your Project:**

   Once connected to the remote machine, open your project folder or the source code you want to debug. You can use the integrated terminal to navigate to the project directory.

5. **Create a Launch Configuration for Debugging:**

   Depending on the programming language you're using, you need to set up a debug configuration. For C/C++, you can create a `launch.json` file in your project's `.vscode` folder. Here's an example configuration that includes running the program with `sudo`:

   ```json
   {
       "version": "0.2.0",
       "configurations": [
           {
               "name": "Debug with sudo",
               "type": "cppdbg",
               "request": "launch",
               "program": "${workspaceFolder}/your_program",
               "args": [],
               "stopAtEntry": false,
               "cwd": "${workspaceFolder}",
               "environment": [],
               "externalConsole": false,
               "MIMode": "gdb",
               "setupCommands": [
                   {
                       "description": "Enable pretty-printing for gdb",
                       "text": "-enable-pretty-printing",
                       "ignoreFailures": true
                   }
               ],
               "preLaunchTask": "sudo_debug",
               "miDebuggerPath": "/usr/bin/gdb"
           }
       ]
   }
   ```

   Make sure to replace `${workspaceFolder}/your_program` with the path to your compiled program.

6. **Configure a Pre-Launch Task:**

   In the `launch.json` configuration above, there's a pre-launch task named "sudo_debug." You can define this task in your project's `.vscode/tasks.json` file. Here's an example:

   ```json
   {
       "version": "2.0.0",
       "tasks": [
           {
               "label": "sudo_debug",
               "command": "sudo",
               "args": ["-E", "${workspaceFolder}/your_program"],
               "type": "shell",
               "group": {
                   "kind": "build",
                   "isDefault": true
               }
           }
       ]
   }
   ```

   Replace `${workspaceFolder}/your_program` with the correct path to your program.

7. **Start Debugging:**

   Now, you can start debugging by selecting the "Debug with sudo" configuration from the Run and Debug sidebar. VS Code will launch your program with `sudo` and attach the debugger.

Make sure to adapt the above steps to your specific project and needs. This setup should allow you to debug a Linux target that requires `sudo` to run from within VS Code.

**Comment by gcampbell-msft** [comment]
@tsweaver321 Thanks for following up. I agree that the ChatGPT suggestion seems reasonable. I do believe that anything that requires `sudo` relies more on the machine connection / configuration, and I don't think there is anything explicitly that a launch.json can do, besides running some specific task that in some way activates sudo permissions. I believe that steps 2 and 3 point to this truth as well.

There are also some suggestions that might be useful here: https://stackoverflow.com/questions/40033311/how-to-debug-programs-with-sudo-in-vscode. 

Could you try these out and let us know if they address your issue? 

Additionally, as it relates to debugging, this is more dependent on the vscode-cpptools extension than this extension. @sean-mcmanus Should this issue be transferred?

**Comment by tsweaver321a** [comment]
I hadn't stated it but the project I'm referring to is C/C++.  I previously had attempted to add "sudo": true to my launch.json and it didn't work.  I'll give the ChatGPT solution a try and report back in the near future.  

Thanks for following up on this.

**Comment by andreeis** [comment]
@tsweaver321 and additionally to the solution provided by ChatGPT, there is still opportunity to leverage Makefile Tools for your convenience. 
"args": ["-E", "${workspaceFolder}/your_program"] 
can be
"args": ["-E", "${command:makefile.launchTargetPath}"] 
or ${command:makefile.getLaunchTargetPath}" if you don't want a build to be triggered (the above triggers a build to make sure the binary will be up to date)

We have workitem https://github.com/microsoft/vscode-makefile-tools/issues/382 to track a closer parity between launch.json and "makefile.launchConfigurations" so in future you should be able to use "sudo" via the extension debugging workflow.

I don't think this needs to be transfered to cpptools. @tsweaver321 should confirm if using launch.json with the Makefile Tools command variables work. Then we can close. The other part we need to do is work item 382.
